---
title: "McShane et al. Direct RNA nanopore analyses"
output: html_document
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}

knitr::opts_chunk$set(message = FALSE,        # Hide messages/warnings/errors from loading packages
                      warning = FALSE,
                      error = FALSE,
                      cache = TRUE)           # By default, cache results to avoid re-running
                                              # super long things every time you knit
                      
```


```{r, include=FALSE}
library(ggplot2)
library(dplyr)
library(tidyr)
library(RColorBrewer)
library(viridis)
library(ggpubr)
library(mosaic)
library(biomaRt)
library(gplots)
library(ggbeeswarm)

'%!in%' <- function(x,y)!('%in%'(x,y))

getPalette <-  colorRampPalette(brewer.pal(9, "Spectral"))
cbPalette <- getPalette(12)

getPalette1 <- colorRampPalette(brewer.pal(8, "Dark2"))
getPalette2 <- colorRampPalette(brewer.pal(8, "Set2"))
cbPalette_long <- c(getPalette1(8), getPalette2(8))

ordered_mito_gene_list = c('MT-ND1','MT-ND2','MT-CO1','MT-CO2','MT-ATP8-6','MT-CO3','MT-ND3','MT-ND4L-4','MT-ND5','MT-CYB')

```


## 5'-end processing in polyA+ samples

```{r}

# Get processing data that was computed on HPC cluster
myo_D0_5prime = read.table("/path/to/myoblasts_D0_polyA_enriched_processed_vs_unprocessed_counts_mito_5prime.txt", sep="\t", header=T)
myo_D7_5prime = read.table("/path/to/myoblasts_D7_polyA_enriched_processed_vs_unprocessed_counts_mito_5prime.txt", sep="\t", header=T)
K562_1_5prime = read.table("/path/to/K562_rep1_processed_vs_unprocessed_counts_mito_5prime.txt", sep="\t", header=T)
K562_2_5prime = read.table("/path/to/K562_rep2_processed_vs_unprocessed_counts_mito_5prime.txt", sep="\t", header=T)
HeLa_5prime = read.table("/path/to/HeLa_polyA+_processed_vs_unprocessed_counts_mito_5prime.txt", sep="\t", header=T)

# Add sample name
myo_D0_5prime$sample_name <- "myoblasts_Day0"
myo_D7_5prime$sample_name <- "myoblasts_Day7"
K562_1_5prime$sample_name <- "K562_rep1"
K562_2_5prime$sample_name <- "K562_rep2"
HeLa_5prime$sample_name <- "HeLa"

# Concatenate all samples and calculate proportion of processed reads
df_5prime <- rbind(myo_D0_5prime,myo_D7_5prime,K562_1_5prime,K562_2_5prime,HeLa_5prime)

df_5prime <- df_5prime %>% mutate(ratio = unprocessed_5prime / (processed_5prime + unprocessed_5prime)) %>% filter(gene_name != "MT-ND6")
df_5prime$gene_name <- factor(df_5prime$gene_name, levels=ordered_mito_gene_list)
df_5prime$sample_name <- factor(df_5prime$sample_name, levels=c("HeLa", "K562_rep1","K562_rep2","myoblasts_Day0","myoblasts_Day7"))


df_5prime %>%
  ggplot(aes(x=gene_name, y=ratio, fill=sample_name)) + geom_bar(stat="identity", position="dodge") + theme_bw() + scale_fill_manual(values=c(cbPalette_long[2],cbPalette_long[3],cbPalette_long[3],cbPalette_long[1],cbPalette_long[5])) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ylab("fraction of unprocessed reads") + ylim(0,1)


ggsave("/path/to/all_cell_lines_with_HeLa_5prime_processing_proportions.pdf",
       width = 6,
    height = 2.5,
    units = c("in"))

```




## 5'-end processing vs. poly(A) tail length

```{r}

# Load data that was produced on HPC cluster
## Per-read processing level
myo_D0_process_df = read.table("/path/to/myoblasts_D0_polyA_enriched_read_processing_status.txt", sep="\t", header=T)
myo_D7_process_df = read.table("/path/to/myoblasts_D7_polyA_enriched_read_processing_status.txt", sep="\t", header=T)
K562_1_process_df = read.table("/path/to/K562_rep1_read_processing_status.txt", sep="\t", header=T)
K562_2_process_df = read.table("/path/to/K562_rep2_read_processing_status.txt", sep="\t", header=T)
HeLa_process_df = read.table("/path/to/HeLa_polyA+_read_processing_status.txt", sep="\t", header=T)

## Nanopolish poly(A) tail estimates
myo_D0_nano_df = read.table("/path/to/myoblasts_D0_polyA_enriched_polyA_estimates.tsv", sep="\t", header=T)
myo_D7_nano_df = read.table("/path/to/myoblasts_D7_polyA_enriched_polyA_estimates.tsv", sep="\t", header=T)
K562_1_nano_df = read.table("/path/to/K562_rep1_polyA_enriched_polyA_estimates.tsv", sep="\t", header=T)
K562_2_nano_df = read.table("/path/to/K562_rep2_polyA_enriched_polyA_estimates.tsv", sep="\t", header=T)
HeLa_nano_df = read.table("/path/to/HeLa_polyA+_polyA_enriched_polyA_estimates.tsv", sep="\t", header=T)

myo_D0_nano_df$sample_name <- "myoblasts_Day0"
myo_D7_nano_df$sample_name <- "myoblasts_Day7"
K562_1_nano_df$sample_name <- "K562_rep1"
K562_2_nano_df$sample_name <- "K562_rep2"
HeLa_nano_df$sample_name <- "HeLa"

process_df <- rbind(myo_D0_process_df, myo_D7_process_df, K562_1_process_df, K562_2_process_df, HeLa_process_df)
nano_df <- rbind(myo_D0_nano_df, myo_D7_nano_df, K562_1_nano_df, K562_2_nano_df, HeLa_nano_df) %>% 
  filter(qc_tag %in% c("PASS","ADAPTER") & grepl("MT", contig))

df <- inner_join(process_df, nano_df, by=c("read"="readname")) %>% filter(category_5prime != "other") %>%
  dplyr::select(read, gene_name, polya_length,category_5prime, category_3prime, sample_name) %>%
  filter(category_3prime == "processed_3prime" & !grepl("RNR", gene_name) & !grepl("MT-T", gene_name) & !grepl("anti", gene_name) & gene_name != "MT-ND6") %>% distinct()

ordered_mito_gene_list = c('MT-ND1','MT-ND2','MT-CO1','MT-CO2','MT-ATP8-6','MT-CO3','MT-ND3','MT-ND4L-4','MT-ND5','MT-CYB')

df$gene_name <- factor(df$gene_name, levels=ordered_mito_gene_list)

df$sample_name <- factor(df$sample_name, levels=c("HeLa", "K562_rep1","K562_rep2","myoblasts_Day0","myoblasts_Day7"))


```


## Plot poly(A) tail lengths in HeLa cells

```{r}

df %>% filter(sample_name == "HeLa") %>%
  ggplot(aes(x=gene_name, y=polya_length)) + geom_violin() + geom_boxplot(outlier.shape = NA) + theme_bw() + ylim(0,120) +
  scale_colour_manual(values=c(cbPalette_long)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


ggsave("/path/to/mito_polyA_length_HeLa_violin_and_boxplot.pdf",
       width = 6,
    height = 3,
    units = c("in"))


```

## Plot poly(A) tail lengths in all cell types

```{r}

df %>%
  ggplot(aes(x=gene_name, y=polya_length, fill=sample_name)) + geom_boxplot(outlier.shape = NA) + theme_bw() + ylim(0,120) +
  scale_fill_manual(values=c(cbPalette_long[2],cbPalette_long[3],cbPalette_long[3],cbPalette_long[1],cbPalette_long[5])) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


ggsave("/path/to/mito_polyA_length_K562_myoblasts_HeLa_boxplot.pdf",
       width = 5,
    height = 2.5,
    units = c("in"))


```


## Poly(A) tail length vs. 5'-end processing status

```{r}

df %>% filter(sample_name == "HeLa" & gene_name %in% c("MT-CO1","MT-ND1","MT-CYB","MT-CO3")) %>% filter(category_5prime != "other") %>%
  ggplot(aes(x=polya_length, colour=category_5prime)) + geom_density() + theme_bw() + xlim(0,100) +
  scale_colour_manual(values=cbPalette_long) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + facet_wrap(~gene_name, ncol=4)

ggsave("/path/to/mito_polyA_length_processed_vs_unprocessed_HeLa_density_plot.pdf",
       width = 8,
    height = 2,
   units = c("in"))

```


```{r}
# Calculate read counts per category
df %>% filter(sample_name == "HeLa" & gene_name %in% c("MT-CO1","MT-ND1","MT-CYB","MT-CO3")) %>% filter(category_5prime != "other") %>%
  group_by(gene_name, category_5prime) %>% summarise(n=n())


```


```{r}
# Calculate difference in processed/unprocessed per transcript
df %>% filter(sample_name == "HeLa" & gene_name %in% c("MT-CO1","MT-ND1","MT-CYB","MT-CO3")) %>% filter(category_5prime != "other") %>%
  group_by(gene_name, category_5prime) %>% summarise(median_polya=median(polya_length)) %>% spread(category_5prime, median_polya) %>%
  mutate(delta = processed_5prime - unprocessed_5prime)


```

```{r}
# Stats
df %>% filter(gene_name %in% c("MT-CO1","MT-ND1","MT-CYB","MT-CO3")) %>% filter(category_5prime != "other") %>%
  group_by(gene_name, sample_name) %>% do(w = wilcox.test(polya_length~category_5prime, data=., paired=FALSE)) %>%
  summarise(gene_name, sample_name, Wilcox = w$p.value) %>% filter(Wilcox < 0.05)


```



## Make Supplemental Tables

```{r}

supp_table_polyA <- df %>% group_by(sample_name, gene_name) %>% summarise(median_polyA_length = median(polya_length))


write.table(supp_table_polyA, "/path/to/Supp_Table_median_polyA_length_all_polyA_enriched_samples.txt", sep="\t", row.names=F, col.names=T, quote=F)

```


```{r}

supp_table_5prime_polyA <- df %>% filter(gene_name %in% c("MT-CO1","MT-ND1","MT-CYB","MT-CO3")) %>% filter(category_5prime != "other") %>%
  group_by(sample_name, gene_name, category_5prime) %>% summarise(median=median(polya_length)) %>% spread(category_5prime, median)


write.table(supp_table_5prime_polyA, "/path/to/Supp_Table_median_polyA_length_vs_5prime_processing_all_polyA_enriched_samples.txt", sep="\t", row.names=F, col.names=T, quote=F)

```


